### PR TITLE
chore(remap): Make del function only take one arg

### LINKED ISSUE #5605: remove multiple arguments from `del` function
We want to handle this using built-in iteration support at some point, so let's remove multiple arguments from this function for now. Also, we wanted this function to return the value of the field being deleted, such that you can do:

```rust
.ts = to_timestamp(del(.timestamp))
```

So we should make that change when we remove multiple-argument support as well.

cc: @binarylogic @FungusHumungus

### PR DESCRIPTION
**Pull request by lucperkins**
At the moment, the `del` function can be used to delete any number of object fields. This capability, however, is inconsistent with an emerging core tenet of TRL, which is that functions should return a single value. If `del` is called on two fields and only one exists, it's not clear what the appropriate response should be. Furthermore, it's been decided that `del` should return the value of the field deleted rather than `null`, in order to "bubble up" the value of the field upon deletion and thus enable expressions like this:

```
.ts = to_timestamp(del(.timestamp))
```

Closes #5605

### CODE REVIEW

**Comment by lucperkins** [comment]
@JeanMertz @FungusHumungus There's a bit more work that needs to be done here to have `del` return the value, but I'm not sure about the desired behavior. If field `timestamp` doesn't exist and I run `del(.timestamp)`, should that return `null` or an error?

**Comment by binarylogic** [comment]
@lucperkins does this close an issue?

**Comment by lucperkins** [comment]
@binarylogic It *might* address #5566, but I'm going to hold off on pushing any further on this until there's consensus around expected behavior.

**Comment by jamtur01** [comment]
Also linked to #5605.

**Comment by jamtur01** [comment]
To move this forward: @binarylogic @lucperkins For this release we just need to answer: should calling `del` on a non-existing field error or return `null`?

**Comment by binarylogic** [comment]
For now, it should return `null` until we finalize the error handling story.

**Comment by FungusHumungus** [comment]
It looks like this is always returning `Null`. It would be really handy if we could address #5150 here as well so that `del` returns the value of the field being deleted - only returning `Null` if the field doesn't exist.

I'm sure I saw somewhere you added a `remove_and_get` function to `Object` that handled this case, but I can't seem to find it now.

**Comment by lucperkins** [comment]
@FungusHumungus You're not just seeing things 😄 I did have that behavior here but deferred to @binarylogic on returning `null` for now, which I interpreted as *always* return `null` but on a second pass I'm reading this as return `null` only if the field is missing. Ben, could you clarify which of the following you prefer:

1. `del` always returns `null`
2. `del` returns the value of the field if it exists and `null` if it doesn't

**Comment by lucperkins** [comment]
@FungusHumungus Okay, I've reverted that behavior. I've left `remove_and_get` unimplemented for now in `LogEvent` and `Metric`. As far as I can tell those implementations aren't currently necessary for VRL, but if that's incorrect, let me know and I'll dig in and supply those impls.

**Comment by FungusHumungus** [comment]
> As far as I can tell those implementations aren't currently necessary for VRL, but if that's incorrect, let me know and I'll dig in and supply them, though I may need a bit of help.

They will be necessary.

I think the hardest things will be if you `del(.)` for a log event, it will need to iterate over the fields to build up a `BTreeMap` of the values before deleting and then return `map.into()`. 

Metric events will just error if you `del(.)`, but the other fields you need to create a Value from the deleted fields.

I'm here if you need a hand doing it, or if you prefer I can have a bash at them and push the code here..

File: docs/reference/remap/del.cue [hunk 1]
```diff
@@ -3,28 +3,28 @@ package metadata
...
-	return: ["null"]
+	return: ["any"]
 	category: "Event"
 	description: #"""
-		Removed the fields specified by the given paths from the root `event` object. Multiple fields can be specified.
+		Removed the field specified by the given path from the event object. If the field exists,
+		the field's value is returned by the delete operation; if the field does not exist, `null`
+		is returned.
 		"""#
 	examples: [
```
Comment (FungusHumungus) [inline docs/reference/remap/del.cue:19]: It would be nice to have another example showing the renaming of a field here..

`.field2 = del(.field1)`

**Comment by lucperkins** [comment]
@FungusHumungus I poked around with providing the `LogEvent` impl (`Metric` is easy, of course) and am confident I could make it happen with some effort but I suspect you'd make very short work of it. So if you have some extra time to push a solution to this branch, that'd be by far the most efficient way to go about it.

**Comment by FungusHumungus** [comment]
@lucperkins Have pushed a commit that should do this.

File: docs/reference/remap/append.cue [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package metadata
```
Comment (JeanMertz) [inline docs/reference/remap/append.cue:1]: I guess this was accidentally added?

File: docs/reference/remap/del.cue [hunk 3]
```diff
@@ -3,28 +3,28 @@ package metadata
...
 		},
 	]
-	return: ["null"]
+	return: ["any"]
 	category: "Event"
 	description: #"""
-		Removed the fields specified by the given paths from the root `event` object. Multiple fields can be specified.
+		Removed the field specified by the given path from the event object. If the field exists,
+		the field's value is returned by the delete operation; if the field does not exist, `null`
+		is returned.
```
Comment (JeanMertz) [inline docs/reference/remap/del.cue:17]: I'm not sure where I proposed this, but I proposed that we return an _error_ if the field does not exist.

That is the only way we can distinguish between a field that exists and has a `null` value, and a field that doesn't exist.

File: lib/remap-functions/src/del.rs [hunk 4]
```diff
@@ -9,47 +9,54 @@ impl Function for Del {
...
 
-        Ok(Box::new(DelFn { paths }))
+        Ok(Box::new(DelFn { field }))
     }
 }
 
 #[derive(Debug, Clone)]
 pub struct DelFn {
-    paths: Vec<Path>,
+    field: Path,
```
Comment (JeanMertz) [inline lib/remap-functions/src/del.rs:28]: Let's use `path` instead of `field`, since that's what we're storing here.

File: lib/remap-functions/src/del.rs [hunk 5]
```diff
@@ -9,47 +9,54 @@ impl Function for Del {
...
 }
 
 impl Expression for DelFn {
     fn execute(&self, _: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
-        self.paths
-            .iter()
-            .try_for_each(|path| object.remove(path.as_ref(), false))?;
-
-        Ok(Value::Null)
+        match object.remove_and_get(self.field.as_ref(), false) {
```
Comment (JeanMertz) [inline lib/remap-functions/src/del.rs:40]: We should probably just update `remove` to always return the removed value.

You can always ignore that value if it's not of use.

File: lib/remap-functions/src/del.rs [hunk 6]
```diff
@@ -9,47 +9,54 @@ impl Function for Del {
...
+            fallible: false,
+            kind: Kind::Bytes
+                | Kind::Integer
+                | Kind::Float
+                | Kind::Boolean
+                | Kind::Map
+                | Kind::Array
+                | Kind::Timestamp
+                | Kind::Regex
+                | Kind::Null,
```
Comment (JeanMertz) [inline lib/remap-functions/src/del.rs:59]: There's a `Kind::all()` so we don't have to keep updating this if we add new value kinds.

**Review by JeanMertz — CHANGES_REQUESTED** [review]
This is looking good, but isn't quite what we want just yet. I left some notes on what I think we should do before we merge this PR.

File: docs/reference/remap/del.cue (hunk 3 above)
Comment (JeanMertz) [reply docs/reference/remap/del.cue:17]: Here's the issue @binarylogic created based on my earlier comment: https://github.com/timberio/vector/issues/5779

Comment (FungusHumungus) [reply docs/reference/remap/del.cue:17]: We decided to just return Null for now and revisit it again once we have fully nailed down the error handling. See comment from @binarylogic above...

*(Hope you had a good holiday!)*

File: docs/reference/remap/append.cue (hunk 2 above)
Comment (lucperkins) [reply docs/reference/remap/append.cue:1]: No, this is present in every file under `docs/reference`.

File: docs/reference/remap/del.cue (hunk 3 above)
Comment (JeanMertz) [reply docs/reference/remap/del.cue:17]: > (Hope you had a good holiday!)

I had, thanks @FungusHumungus 😄 

As for the decision to return `null`, I don't think that's a good idea.

For one, it means you can't reliably use `del` to return the value of the deleted field, since you don't know if the field contained null or didn't exist. Second, we want to resolve error handling before we ship the first version regardless, so we might as well move ahead and make this function fallible while we also continue to work on the error handling story.

File: lib/remap-lang/src/object.rs [hunk 7]
```diff
@@ -56,4 +56,10 @@ pub trait Object: std::fmt::Debug {
     /// If `compact` is true, after deletion, if an empty object or array is
     /// left behind, it should be removed as well.
     fn remove(&mut self, path: &Path, compact: bool) -> Result<(), String>;
+
+    /// Remove the given path from the object and return the value if it exists.
+    ///
+    /// If `compact` is true, after deletion, if an empty object or array is
+    /// left behind, it should be removed as well.
+    fn remove_and_get(&mut self, path: &Path, compact: bool) -> Result<Option<Value>, String>;
```
Comment (JeanMertz) [inline lib/remap-lang/src/object.rs:64]: Let's remove this addition, and update

```rust
fn remove(&mut self, path: &Path, compact: bool) -> Result<(), String>;
```

to:

```rust
fn remove(&mut self, path: &Path, compact: bool) -> Result<Value, String>;
```

**Review by JeanMertz — CHANGES_REQUESTED** [review]

File: docs/reference/remap/del.cue (hunk 3 above)
Comment (JeanMertz) [reply docs/reference/remap/del.cue]: So, I've seen the example given in the announcement post:

```rust
ok(del(.user.id))
ok(del(.user.ip_address))
.log_stream_name = "system_logs"
```

I've also since this comment written about our [final error handling story](https://gist.github.com/JeanMertz/f36e0aae6191231d1a37d81ba560047e). Specifically, I've written about allowing functions to [define their fallibility](https://gist.github.com/JeanMertz/f36e0aae6191231d1a37d81ba560047e#per-function-ignore-error-options) based on parameters.

Given the new context and the provided example, I've come around to my earlier conclusion.

Let's keep the above example, and go through the different possibilities:

1. Use the `!` bang-function to abort.

   ```rust
   del!(.user.id)
   del!(.user.ip_address)
   .log_stream_name = "system_logs"
   ```

   This works, but I can't imagine you want to abort the program here because if the field didn't exist, you don't mind if it wasn't deleted anyway. The end-result is the same, regardless if the `del` function call succeeded.

2. Use the `?` error coalesce operator to silence errors.

   ```rust
   del(.user.id) ? null
   del(.user.ip_address) ? null
   .log_stream_name = "system_logs"
   ```

   This looks weird, since you're not doing anything with the `null` value, so it's basically a way to hide the error.

3. Use the error-assignment technique to handle/ignore errors.

   ```rust
   _, _ = del(.user.id)
   _, _ = del(.user.ip_address)
   .log_stream_name = "system_logs"
   ```

   This has the same problem as the error coalesce operator except worse, there's a lot of ugly syntax here just to hide an error because we don't care about both the error and the return value in this case.

4. Use the `ignore_errors` parameter to silence errors.

   ```rust
   del(.user.id, true)
   del(.user.ip_address, true)
   .log_stream_name = "system_logs"
   ```

   I omitted the `ignore_errors` parameter keyword for brevity, but even so, this is more verbose than you'd like for something you'll use as frequent as the `del` function.

Given the above examples, I'm inclined to go with (4), but flip the defaults:

If the second argument is `false` (or undefined), then `del` is an infallible function that returns `null` if the field didn't exist, and otherwise returns the value of the field. If it's set to `true`, then the function becomes fallible, and errors if the field didn't exist:

```rust
del(.user.id)
del!(.user.ip_address, true) // this aborts if the field didn't exist
.log_stream_name = "system_logs"
```

As for my reasoning:

The first argument for this proposal is that it is much easier on the eyes, and (I believe) favours the most frequent use-case (don't care about the return value, or more specifically, don't care if I can't distinguish between `null` fields and non-existing fields).

The second argument (which I hadn't thought about until just now) is that it is actually _more consistent_ with other parts of the language. Specifically, if I query a field `.foo.bar`, and that field does not exist, we _also_ return `null` as the default value, so even there you can't know if the field was set and has `null` as its value, or if the field doesn't exist (we provide the `exists` function to make that distinction, which can also be used when calling `del`).

So, in conclusion, my proposal would be:

* have `del` take two arguments: `value` and `error_when_missing`
* `error_when_missing` defaults to `false`
* make `del` infallible by default, and return `null` if the field doesn't exist
* `error_when_missing = true` makes `del` become fallible, returning an error if the field doesn't exist

_(implementation detail note: the `error_when_missing` parameter has to take a _literal boolean_ as its value because we have to determine the value of the boolean at compile-time to be able to define the function as fallible or infallible based on this value)_

Comment (jszwedko) [reply docs/reference/remap/del.cue]: > (implementation detail note: the error_when_missing parameter has to take a literal boolean as its value because we have to determine the value of the boolean at compile-time to be able to define the function as fallible or infallible based on this value)

This caveat is a bit unfortunate. I realize why and that users are unlikely to want to use a variable in this particular case, but it still makes the language feel a bit inconsistent to me. If we really want to pursue this route (see my alternative suggestion below) than I might advocate for making this two separate functions: `del` and `edel` (we can work on the names) to avoid introducing this special case.

I think I'm actually not a fan of having `del()` change "abort" behavior based on a boolean flag as I think this is slightly confusing as a user and complicates the internal implementation by requiring the above caveat about it being a literal boolean. I might again prefer to borrow a pattern from Go and have it simply return two values:

* The value of the deleted path or `null` if the path was undefined
* A boolean indicating whether there was actually anything at that path

Users would call this as:

```
user_id, ok = del(.user.id)
```

They can then check `ok` to see if `user_id` was actually defined.

Alternatively we simply keep it as `user_id = del(.user_id)` and users can check if `.user_id` was defined before calling `del()` if they care (pseudocode):

```
if (is_defined(.user_id)) {
  user_id = del(.user_id)
} else {
  user_id = "unknown"
}
```

Notably if we don't change `del()` to return the value, users can also simply do:

```
user_id = .user_id
del(.user_id)
```

Which I don't think is that bad for the initial release and we could modify `del()`'s behavior later in a compatible fashion.

Comment (JeanMertz) [reply docs/reference/remap/del.cue]: Thanks for the input @jszwedko. I think we diverge a bit on the priorities for the language (which is on me, since I haven't taken the time to flesh out the specification yet in https://github.com/timberio/vector/issues/3740).

I think your proposal veers too much from what we've been doing so far with the language (using separate functions for error/non-error and using multiple return values), but obviously, we can pivot into such a direction if others agree.

Of the two proposals, I'd favour separate functions over multiple return values, but I still favour adding the proposed argument to `del`.

To add a few more points supporting that proposal:

* we already have literal string and array requirements for some arguments, so I don't consider literal booleans to be a problem
* the goal of the proposal/this PR is to make `del` as easy to use as possible, while still allowing "accurate" usage if needed, so while we already support `exists` and if-statements, I don't consider those to be valid alternatives
* in the end, I think we should optimize for the 80% use-case, while providing a low-key solution for the other 20%, which is why I prefer infallible `del` by default, and a flag to make it fallible (but more correct).

I want to avoid this devolving into a lengthy discussion, so if you want, we can discuss this on Discord further to reach consensus.

Comment (JeanMertz) [reply docs/reference/remap/del.cue]: @jszwedko and I discussed this offline. It's always a pleasure to debate things like this with you, Jesse 😄.

Here's the conclusion:

I misread this part of Jesse's comment:

> Alternatively we simply keep it as `user_id = del(.user_id)` and users can check if `.user_id` was defined before calling `del()` if they care (pseudocode):
> 
> ```rust
> if exists(.user.id) {
>     .user_id = del(.user.id)
> } else {
>     .user_id = "unknown"
> }
> ```

Having now understood this part of the proposal better, given the fact that we want to "optimize for the 80% use-case", I agree with this direction.

So in conclusion, what we want to do is:

1. make `del` always infallible
2. have it return the value of the deleted field
3. return `null` if a field doesn't exist
4. document the usage of `exists` for people that want to distinguish between existing `null`, and non-existing fields (minority use-case)

**Comment by JeanMertz** [comment]
I updated the PR based on the latest consensus in https://github.com/timberio/vector/pull/5633#discussion_r552716891. This should be good for a final round of reviews before we merge it.

File: docs/reference/remap.cue [hunk 8]
```diff
@@ -49,7 +49,7 @@ remap: {
 
 			arguments: [...#Argument] // Allow for empty list
 			return: [#RemapReturnTypes, ...#RemapReturnTypes]
-			category:    "Coerce" | "Encode" | "Enumerate" | "Event" | "Hash" | "IP" | "Map" | "Number" | "Parse" | "Random" | "String" | "Test" | "Timestamp"
+			category:    "Array" | "Coerce" | "Encode" | "Enumerate" | "Event" | "Hash" | "IP" | "Map" | "Number" | "Parse" | "Random" | "String" | "Test" | "Timestamp"
```
Comment (jszwedko) [inline docs/reference/remap.cue:52]: Did you intend to include `Array` here?

File: docs/reference/remap/del.cue [hunk 9]
```diff
@@ -3,32 +3,40 @@ package metadata
...
-			source: #"""
-				del(.field1, .field3)
-				"""#
+			source: "del(.field1)"
 			output: {
 				"field2": 2
 			}
 		},
+		{
+			title: "Rename field"
```
Comment (jszwedko) [inline docs/reference/remap/del.cue:32]: Maybe we should add an example showing how you can check if the field exists?

File: lib/remap-functions/src/del.rs [hunk 10]
```diff
@@ -9,47 +9,58 @@ impl Function for Del {
...
+        // fields is infallible, deleting metric fields is not.
+        //
+        // For example, if you try to delete `.name` in a metric event, the
+        // call returns an error, since this is an infallible field.
+        //
+        // After some debating, we've decided to _silently ignore_ deletions
+        // of infallible fields for now, but we'll circle back to this in
+        // the near future to potentially improve this situation.
+        //
+        // see tracking issue: <TODO>
```
Comment (jszwedko) [inline lib/remap-functions/src/del.rs:53]: Could we open the issue and link it here?

File: lib/remap-functions/src/del.rs [hunk 11]
```diff
@@ -58,15 +69,58 @@ impl Expression for DelFn {
...
+            ),
+            (
+                // Map field exists
+                map!["exists": map!["foo": "bar"]],
+                Ok(value!(map!["foo": "bar"])),
+                DelFn::new(Path::from("exists")),
+            ),
+            (
+                // Integer field exists
+                map!["exists": 127],
```
Comment (jszwedko) [inline lib/remap-functions/src/del.rs:109]: Would it make sense to add a test for arrays here too? Like `del(.exists[0])`? It seems like `del()` supports this looking at the below `object_remove` tests. I'm not sure where the line is between what should be tested where though.

**Review by jszwedko — COMMENTED** [review]
Looks good! Just a few small suggestions / questions.

File: lib/remap-functions/src/del.rs (hunk 10 above)
Comment (JeanMertz) [reply lib/remap-functions/src/del.rs:53]: Woops, yes, I'll add a link to https://github.com/timberio/vector/issues/5887

File: lib/remap-functions/src/del.rs (hunk 11 above)
Comment (JeanMertz) [reply lib/remap-functions/src/del.rs:109]: I can add it, seems like something we want to make sure keeps working 👍

File: docs/reference/remap.cue (hunk 8 above)
Comment (JeanMertz) [reply docs/reference/remap.cue:52]: Ah no, I removed a bunch of files that got sucked into this PR due to a bad rebase, but didn't realize this was also part of that bad rebase. Will fix.

File: docs/reference/remap/del.cue [hunk 12]
```diff
@@ -28,6 +28,24 @@ remap: functions: del: {
 				"field2": 2
 			}
 		},
+		{
+			title: "Delete existing only"
+			input: {
+				"user_id": 1
+			}
+			source: #"""
+				.user.id = if exists(.user_id) {
```
Comment (jszwedko) [inline docs/reference/remap/del.cue:37]: TIL we support `if`s as expressions :smile:

**Review by jszwedko — APPROVED** [review]
Nice work!

**Review by JeanMertz — APPROVED** [review]
